<a href="https://colab.research.google.com/github/Luke-Tetteh/Introduction-to-Python-for-Machine-Learning/blob/second/Introduction_To_Python_for_Machine_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [2]:
# Question 1: How do you create an identity matrix with python?
np.identity(3) #Answer


array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [3]:
# Question 2: For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

url ='https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv'
fuel_data = pd.read_csv(url, error_bad_lines = False)
fuel_data.describe(include = 'all')

record_id  ...  fuel_cost_per_mmbtu
count                      29523  ...         29523.000000
unique                     29523  ...                  NaN
top     f1_fuel_2017_12_195_1_10  ...                  NaN
freq                           1  ...                  NaN
mean                         NaN  ...            19.304354
std                          NaN  ...          2091.540939
min                          NaN  ...           -41.501000
25%                          NaN  ...             1.940000
50%                          NaN  ...             4.127000
75%                          NaN  ...             7.745000
max                          NaN  ...        359278.000000

[11 rows x 11 columns]

In [4]:
fuel_data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [5]:
# Use groupby to count the number of each unique value in the fuel unit column
fuel_data.groupby('fuel_unit')['fuel_unit'].count()

fuel_unit
bbl        7998
gal          84
gramsU      464
kgU         110
mcf       11354
mmbtu       180
mwdth        95
mwhth       100
ton        8958
Name: fuel_unit, dtype: int64

In [6]:
fuel_data[['fuel_unit']] = fuel_data[['fuel_unit']].fillna(value = 'mcf')

In [7]:
# Check if missing values have been filled

fuel_data.isnull().sum()

record_id                       0
utility_id_ferc1                0
report_year                     0
plant_name_ferc1                0
fuel_type_code_pudl             0
fuel_unit                       0
fuel_qty_burned                 0
fuel_mmbtu_per_unit             0
fuel_cost_per_unit_burned       0
fuel_cost_per_unit_delivered    0
fuel_cost_per_mmbtu             0
dtype: int64

In [8]:
fuel_data.describe()

utility_id_ferc1  ...  fuel_cost_per_mmbtu
count      29523.000000  ...         29523.000000
mean         118.601836  ...            19.304354
std           74.178353  ...          2091.540939
min            1.000000  ...           -41.501000
25%           55.000000  ...             1.940000
50%          122.000000  ...             4.127000
75%          176.000000  ...             7.745000
max          514.000000  ...        359278.000000

[8 rows x 7 columns]

In [16]:
fuel_count = fuel_data.groupby('report_year').count()
fuel_count

record_id  ...  fuel_cost_per_mmbtu
report_year             ...                     
1994              1235  ...                 1235
1995              1201  ...                 1201
1996              1088  ...                 1088
1997              1094  ...                 1094
1998              1107  ...                 1107
1999              1050  ...                 1050
2000              1373  ...                 1373
2001              1356  ...                 1356
2002              1205  ...                 1205
2003              1211  ...                 1211
2004              1192  ...                 1192
2005              1269  ...                 1269
2006              1243  ...                 1243
2007              1264  ...                 1264
2008              1228  ...                 1228
2009              1222  ...                 1222
2010              1261  ...                 1261
2011              1240  ...                 1240
2012              1243  ...                 1243
2013              1199  ...                 1199
2014              1171  ...                 1171
2015              1093  ...                 1093
2016              1034  ...                 1034
2017               993  ...                  993
2018               951  ...                  951

[25 rows x 10 columns]

In [10]:
fuel.pct_change(periods=4)

record_id  ...  fuel_cost_per_mmbtu
report_year             ...                     
1994               NaN  ...                  NaN
1995               NaN  ...                  NaN
1996               NaN  ...                  NaN
1997               NaN  ...                  NaN
1998         -0.103644  ...            -0.103644
1999         -0.125729  ...            -0.125729
2000          0.261949  ...             0.261949
2001          0.239488  ...             0.239488
2002          0.088528  ...             0.088528
2003          0.153333  ...             0.153333
2004         -0.131828  ...            -0.131828
2005         -0.064159  ...            -0.064159
2006          0.031535  ...             0.031535
2007          0.043765  ...             0.043765
2008          0.030201  ...             0.030201
2009         -0.037037  ...            -0.037037
2010          0.014481  ...             0.014481
2011         -0.018987  ...            -0.018987
2012          0.012215  ...             0.012215
2013         -0.018822  ...            -0.018822
2014         -0.071372  ...            -0.071372
2015         -0.118548  ...            -0.118548
2016         -0.168142  ...            -0.168142
2017         -0.171810  ...            -0.171810
2018         -0.187874  ...            -0.187874

[25 rows x 10 columns]

The percentage change in the fuel cost per unit burned in 1998 compared to 1994 is -0.103644 or about -10.36%


In [11]:
# Question 3: Which of the features has the second and third lowest correlation with the Fuel Cost Per Unit Burned?

In [12]:
fuel_data.corr()

utility_id_ferc1  ...  fuel_cost_per_mmbtu
utility_id_ferc1                      1.000000  ...             0.006122
report_year                           0.093323  ...             0.010261
fuel_qty_burned                      -0.057447  ...            -0.001896
fuel_mmbtu_per_unit                  -0.066946  ...            -0.005884
fuel_cost_per_unit_burned            -0.037863  ...            -0.000437
fuel_cost_per_unit_delivered         -0.016414  ...            -0.000109
fuel_cost_per_mmbtu                   0.006122  ...             1.000000

[7 rows x 7 columns]

Answer: fuel_qty_burned(-0.018535) and fuel_mmbtu_per_unit(-0.000437)

In [13]:
# Question 5: Which year has the highest average fuel cost per unit delivered?


In [19]:
fuel_average = fuel_data.groupby('report_year').mean()
fuel_average

utility_id_ferc1  ...  fuel_cost_per_mmbtu
report_year                    ...                     
1994               106.120648  ...             3.086977
1995               110.139051  ...             2.868838
1996               108.714154  ...             2.695880
1997               108.366545  ...             3.113486
1998               109.845528  ...             2.775862
1999               108.222857  ...             2.740251
2000               111.083758  ...             4.208403
2001               113.521386  ...             4.612658
2002               120.354357  ...             6.226281
2003               123.300578  ...             7.223933
2004               121.123322  ...             8.441597
2005               121.028369  ...             9.397078
2006               117.178600  ...            10.339898
2007               116.040348  ...            10.735181
2008               120.785016  ...            12.690115
2009               118.139116  ...            10.619376
2010               117.992863  ...             8.551217
2011               121.069355  ...             9.237591
2012               120.769912  ...             8.808904
2013               122.082569  ...             9.308225
2014               129.263877  ...            10.064673
2015               128.591949  ...             8.054361
2016               133.117021  ...             6.915613
2017               131.126888  ...           368.736351
2018               133.669821  ...             7.233643

[25 rows x 6 columns]

In [52]:
fuel_average['fuel_cost_per_unit_delivered'].idxmax(axis = 1)

1997

Answer: 1997 is the year with the highest average fuel cost per unit delivered